nemo/collections/asr/parts/utils/streaming_utils.py

```diff
@@ -343,6 +343,129 @@ def inplace_buffer_merge(buffer, data, timesteps, model):
+                Size of the total audio in seconds maintained in the buffer
+            chunk_look_back (float):
+                Size of the look-back length for middle chunk
+        '''
+
+        self.NORM_CONSTANT = 1e-5
+        self.ZERO_LEVEL_SPEC_DB_VAL = -16.635  # Log-Melspectrogram value for zero signal
+        self.asr_model = asr_model
+        self.sr = asr_model._cfg.sample_rate
+        self.chunk_size = chunk_size
+        self.n_chunk_look_back = int(chunk_look_back * self.sr)
+        timestep_duration = asr_model._cfg.preprocessor.window_stride
+        self.n_chunk_samples = int(chunk_size * self.sr)
+        self.buffer_size = buffer_size
+        total_buffer_len = int(buffer_size / timestep_duration)
+        self.n_feat = asr_model._cfg.preprocessor.features
+        self.sample_buffer = np.zeros(int(buffer_size * self.sr))
+        self.buffer = np.ones([self.n_feat, total_buffer_len], dtype=np.float32) * self.ZERO_LEVEL_SPEC_DB_VAL
+        self.feature_chunk_len = int(chunk_size / timestep_duration)
+
+        self.feature_buffer_len = total_buffer_len
+
+        self.feature_buffer = (
+            np.ones([self.n_feat, self.feature_buffer_len], dtype=np.float32) * self.ZERO_LEVEL_SPEC_DB_VAL
+        )
+        self.reset()
+        self.buffered_len = 0
+        cfg = copy.deepcopy(asr_model._cfg)
+        OmegaConf.set_struct(cfg.preprocessor, False)
+
+        cfg.preprocessor.dither = 0.0
+        cfg.preprocessor.pad_to = 0
+        cfg.preprocessor.normalize = "None"
+        self.raw_preprocessor = EncDecCTCModelBPE.from_config_dict(cfg.preprocessor)
+        self.raw_preprocessor.to(asr_model.device)
+
+    def reset(self):
+        '''
+        Reset frame_history and decoder's state
+        '''
+        self.buffer = np.ones(shape=self.buffer.shape, dtype=np.float32) * self.ZERO_LEVEL_SPEC_DB_VAL
+        self.frame_buffers = []
+        self.sample_buffer = np.zeros(int(self.buffer_size * self.sr))
+        self.feature_buffer = (
+            np.ones([self.n_feat, self.feature_buffer_len], dtype=np.float32) * self.ZERO_LEVEL_SPEC_DB_VAL
+        )
+
+    def _add_chunk_to_buffer(self, chunk):
+        """
+        Add time-series audio signal to `sample_buffer`
+
+        Args:
+            chunk (np.array):
+                Numpy array filled with time-series audio signal
+        """
+        self.sample_buffer[: -self.n_chunk_samples] = self.sample_buffer[self.n_chunk_samples :]
+        self.sample_buffer[-self.n_chunk_samples :] = chunk
+
+    def _update_feature_buffer(self, feat_chunk):
+        """
+        Add an extracted feature to `feature_buffer`
+        """
+        self.feature_buffer[:, : -self.feature_chunk_len] = self.feature_buffer[:, self.feature_chunk_len :]
+        self.feature_buffer[:, -self.feature_chunk_len :] = feat_chunk
+
+    def get_raw_feature_buffer(self):
+        return self.feature_buffer
+
+    def get_normalized_feature_buffer(self):
```

Comment:
Isn't it better to use normalize_batch method here?
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/preprocessing/features.py#L59

And pass the normalize_type which is extracted already from the asr_model in the init
.

Reply:
This makes sense. This class is based on another older class we used for streaming simulation. @tango4j I will clean up both the classes later, you can leave this as is for now.

Reply:
I have talked to @jbalam-nv and I converted the whole class to torch.tensor based + used `normalize_batch()` function for consistency.